File: src/10.hs
import Data.List (sort)

import Common (getDayLines)

main = do
  input <- getDayLines 10
  print $ partOne input
  print $ partTwo input

partOne = sum . map score
  where score = go []
        go stack (op:cs) | elem op "([{<" = go (op:stack) cs
        go (op:stack) (cl:cs) | matches op cl = go stack cs
        go _ (cl:_) = points cl
        go _ [] = 0
        points ')' = 3
        points ']' = 57
        points '}' = 1197
        points '>' = 25137

partTwo lines = scores !! (length scores `div` 2)
  where scores = sort $ filter (> 0) $ map score lines
        score = go []
        go stack (op:cs) | elem op "([{<" = go (op:stack) cs
        go (op:stack) (cl:cs) | matches op cl =  go stack cs
        go stack [] = foldl (\acc c -> acc*5 + points c) 0 stack
        go _ _ = 0
        points '(' = 1
        points '[' = 2
        points '{' = 3
        points '<' = 4

matches '(' ')' = True
matches '[' ']' = True
matches '{' '}' = True
matches '<' '>' = True
matches _ _ = False

File: src/10.py
import sys
import fileinput


OPENING = '([{<'
MATCHING = {')': '(', ']': '[', '}': '{', '>': '<'}


def solve_1(lines):
    points = {')': 3, ']': 57, '}': 1197, '>': 25137}
    result = 0
    for line in lines:
        stack = []
        for c in line:
            if c in OPENING:
                stack.append(c)
            else:
                cl = stack.pop()
                if cl != MATCHING[c]:
                    result += points[c]
    return result


def solve_2(lines):
    points = {'(': 1, '[': 2, '{': 3, '<': 4}
    scores = []
    for line in lines:
        stack = []
        is_corrupted = False
        for c in line:
            if c in OPENING:
                stack.append(c)
            else:
                op = stack.pop()
                if op != MATCHING[c]:
                    is_corrupted = True
                    break
        if not is_corrupted:
            score = 0
            for c in reversed(stack):
                score = score*5 + points[c]
            scores.append(score)
    return sorted(scores)[len(scores)//2]


if __name__ == '__main__' and not sys.flags.interactive:
    lines = [line.strip() for line in fileinput.input()]
    print(solve_1(lines))
    print(solve_2(lines))
